### PR TITLE
Feature/continue learning

### PR DESCRIPTION
**Pull request by eadm**
**YouTrack task**: [#APPS-1495](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1495)

**Description List**:
* Change fast continue course widget
* Add widget with current streak

### CODE REVIEW

File: app/src/main/res/layout/fragment_fast_continue.xml [hunk 1]
```diff
@@ -11,66 +12,146 @@
...
-        android:layout_height="wrap_content"
-        android:visibility="gone"
-        tools:visibility="visible"/>
+            <TextView
+                android:id="@+id/fastContinueCourseName"
+                android:layout_width="0dp"
+                android:layout_height="wrap_content"
+                fontPath="fonts/Roboto-Light.ttf"
+                android:gravity="start"
+                android:textSize="16sp"
```
Comment (KirillMakarov) [inline app/src/main/res/layout/fragment_fast_continue.xml]: we have `course_item_titles_text_size`

File: app/src/main/res/layout/fragment_fast_continue.xml [hunk 2]
```diff
@@ -11,66 +12,146 @@
...
+                android:layout_marginLeft="@dimen/fast_continue_info_margin"
+                android:layout_marginStart="@dimen/fast_continue_info_margin"
+                app:layout_constraintHorizontal_weight="1"
+                tools:text="Программирование на python"
+                android:layout_marginRight="@dimen/fast_continue_inner_content_margin"
+                android:layout_marginEnd="@dimen/fast_continue_inner_content_margin"
+                app:layout_constraintRight_toRightOf="parent" />
+
+            <TextView
+                android:id="@+id/fastContinueCourseProgressText"
```
Comment (KirillMakarov) [inline app/src/main/res/layout/fragment_fast_continue.xml]: please, extract to style for this view

File: app/src/main/res/layout/home_streak_view.xml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:padding="12dp"
```
Comment (KirillMakarov) [inline app/src/main/res/layout/home_streak_view.xml]: why padding is `12dp`?

File: app/src/main/res/layout/home_streak_view.xml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        android:layout_toRightOf="@id/streakLightingIcon"
+        android:layout_toEndOf="@id/streakLightingIcon"
+        tools:text="4" />
+
+    <TextView
+        android:id="@+id/streakText"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:textColor="@color/new_accent_color"
+        android:textSize="16sp"
```
Comment (KirillMakarov) [inline app/src/main/res/layout/home_streak_view.xml]: `textColor` is defined in styles

File: app/src/main/res/layout/home_streak_view.xml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        android:layout_centerVertical="true"
+        android:layout_alignParentStart="true"
+        android:layout_alignParentLeft="true" />
+
+    <TextView
+        android:id="@+id/streakCounter"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_centerVertical="true"
+        android:textColor="@color/new_accent_color"
```
Comment (KirillMakarov) [inline app/src/main/res/layout/home_streak_view.xml]: `textColor` is defined in styles

File: app/src/main/res/values-ru/strings.xml [hunk 6]
```diff
@@ -546,4 +546,7 @@
         <item quantity="many">%d курсов</item>
         <item quantity="other">%d курсов</item>
     </plurals>
+    <string name="course_current_progress">Текущий прогресс %d%%</string>
+    <string name="home_streak_counter_prefix">"Вы решаете "</string>
+    <string name="home_streak_counter_suffix">" подряд!"</string>
```
Comment (KirillMakarov) [inline app/src/main/res/values-ru/strings.xml]: create 1 string with placeholder %s

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -556,4 +556,7 @@
         <item quantity="one">%d course</item>
         <item quantity="other">%d courses</item>
     </plurals>
+    <string name="course_current_progress">Current progress: %d%%</string>
+    <string name="home_streak_counter_prefix">"Your current Streak is "</string>
+    <string name="home_streak_counter_suffix">!</string>
```
Comment (KirillMakarov) [inline app/src/main/res/values/strings.xml]: create 1 string with placeholder %s

File: app/src/main/res/layout/home_streak_view.xml (hunk 3 above)
Comment (eadm) [reply app/src/main/res/layout/home_streak_view.xml]: That number from Sketch layout

File: app/src/main/res/values-ru/strings.xml (hunk 6 above)
Comment (eadm) [reply app/src/main/res/values-ru/strings.xml]: `%s` not working with spannable strings

File: app/src/main/java/org/stepic/droid/core/presenters/contracts/HomeStreakView.kt [hunk 8]
```diff
@@ -3,4 +3,5 @@ package org.stepic.droid.core.presenters.contracts
 
 interface HomeStreakView {
     fun showStreak(streak: Int)
+    fun onEmptyStreak()
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/core/presenters/contracts/HomeStreakView.kt]: nice!

File: app/src/main/java/org/stepic/droid/di/home/HomeModule.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import org.stepic.droid.di.qualifiers.BackgroundScheduler
+import org.stepic.droid.di.qualifiers.MainScheduler
+import org.stepic.droid.preferences.SharedPreferenceHelper
+import org.stepic.droid.web.Api
+
+@Module
+class HomeModule {
+    @Provides
+    @HomeScope
+    internal fun provideStreakPresenter(@BackgroundScheduler backgroundScheduler: Scheduler,
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/di/home/HomeModule.kt]: it is not needed, just use `@HomeScope` on `HomeStreakPresenter` class and dagger will resolve it automatically

File: app/src/main/java/org/stepic/droid/ui/custom/RoundedBorderMaskView.kt [hunk 10]
```diff
@@ -14,40 +15,41 @@ constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
...
-
-        if (maskBitmap == null) {
-            updateMask(canvas.width, canvas.height)
-        }
-
-        offscreenCanvas.drawBitmap(maskBitmap, 0f, 0f, maskPaint)
-        canvas.drawBitmap(offscreenBitmap, 0f, 0f, paint)
+        super.draw(bufferCanvas)
+        bufferCanvas?.drawBitmap(maskBitmap, 0f, 0f, maskPaint)
+        canvas.drawBitmap(bufferBitmap, 0f, 0f, null)
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/custom/RoundedBorderMaskView.kt]: Nice! New objects are not created now

File: app/src/main/java/org/stepic/droid/ui/custom/RoundedBorderMaskView.kt [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    private fun updateMask(width: Int, height: Int) {
+        if (width <= 0 || height <= 0) return
+
+        val mask = Bitmap.createBitmap(width, height, Bitmap.Config.ALPHA_8)
+        val canvas = Canvas(mask)
+
+        val paint = Paint(Paint.ANTI_ALIAS_FLAG)
+        val rect = RectF(0f, 0f, width.toFloat(), height.toFloat())
```
[the remark this answers is not in the document]
Comment (KirillMakarov) [reply app/src/main/java/org/stepic/droid/ui/custom/RoundedBorderMaskView.kt]: please, extract `RectF` object to field

File: app/src/main/java/org/stepic/droid/ui/custom/RoundedBorderMaskView.kt [hunk 12]
```diff
@@ -14,40 +15,41 @@ constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
...
+        bufferCanvas = Canvas(bufferBitmap)
 
-        val paint = Paint(Paint.ANTI_ALIAS_FLAG)
         val rect = RectF(0f, 0f, width.toFloat(), height.toFloat())
-        canvas.drawRect(rect, paint)
+        val mask = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
+        val canvas = Canvas(mask)
 
-        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
+        val paint = Paint(Paint.ANTI_ALIAS_FLAG)
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/custom/RoundedBorderMaskView.kt]: please, extract `paint` to field

File: app/src/main/java/org/stepic/droid/ui/fragments/HomeFragment.kt [hunk 13]
```diff
@@ -69,20 +73,21 @@ class HomeFragment : FragmentBase(), HomeStreakView {
...
-            val daysSpannable = SpannableString("$streak $daysPlural")
-            daysSpannable.setSpan(regularFontSpan, 0, daysSpannable.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+        val daysPlural = resources.getQuantityString(R.plurals.day_number, streak)
+        val daysSpannable = SpannableString("$streak $daysPlural")
+        daysSpannable.setSpan(regularFontSpan, 0, daysSpannable.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
 
-            val streakTextSpannableBuilder = SpannableStringBuilder(getString(R.string.home_streak_counter_prefix))
-            streakTextSpannableBuilder.append(daysSpannable)
-            streakTextSpannableBuilder.append(getString(R.string.home_streak_counter_suffix))
+        val streakTextSpannableBuilder = SpannableStringBuilder(getString(R.string.home_streak_counter_prefix))
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/fragments/HomeFragment.kt]: we can call like this
```
val homeStreakStr = getQuantityString() (R.string.home_streak, streak /* it is number */ , streak /* it is format arg*/);
```
and after that create `SpannableString(homeStreakStr)`

File: app/src/main/res/layout/fragment_fast_continue.xml [hunk 14]
```diff
@@ -108,15 +105,11 @@
 
             <TextView
                 android:id="@+id/fastContinueCourseProgressText"
+                style="@style/FastContinueSubtextStyle"
```
Comment (KirillMakarov) [inline app/src/main/res/layout/fragment_fast_continue.xml]: we can extract `textColor`, `gravity` to style too.

File: app/src/main/java/org/stepic/droid/ui/fragments/HomeFragment.kt [hunk 15]
```diff
@@ -76,14 +68,16 @@ class HomeFragment : FragmentBase(), HomeStreakView {
...
-        val streakTextSpannableBuilder = SpannableStringBuilder(getString(R.string.home_streak_counter_prefix))
-        streakTextSpannableBuilder.append(daysSpannable)
-        streakTextSpannableBuilder.append(getString(R.string.home_streak_counter_suffix))
-
-        streakText.text = streakTextSpannableBuilder
+        val days = "$streak $daysPlural"
+
+        streakText.text = getString(R.string.home_streak_counter_text, days).let {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                Html.fromHtml(it, Html.FROM_HTML_MODE_LEGACY)
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/fragments/HomeFragment.kt]: we have `TextResolver`, which do it for us

**Pull request merged**